godot_rl/wrappers/stable_baselines_wrapper.py

```diff
@@ -2,55 +2,109 @@
-    def __init__(self, env_path=None, **kwargs):
-        self.env = GodotEnv(env_path=env_path, convert_action_space=True, **kwargs)
+    def __init__(self, env_path: Optional[str] = None, n_parallel: int = 1, **kwargs) -> None:
+        # If we are doing editor training, n_parallel must be 1
+        if env_path is None and n_parallel > 1:
+            raise ValueError("You must provide the path to a exported game executable if n_parallel > 1")
+        
+        # Define the default port
+        port = kwargs.pop("port", GodotEnv.DEFAULT_PORT)
+
+        # Create a list of GodotEnv instances
+        self.envs = [GodotEnv(env_path=env_path, convert_action_space=True, port=port+p, seed=p, **kwargs) for p in range(n_parallel)]
+        
+        # Store the number of parallel environments
+        self.n_parallel = n_parallel
+
+        # Check the action space for validity
         self._check_valid_action_space()
+
+        # Initialize the results holder
         self.results = None
 
-    def _check_valid_action_space(self):
-        action_space = self.env.action_space
+    def _check_valid_action_space(self) -> None:
+        # Check if the action space is a tuple space with multiple spaces
+        action_space = self.envs[0].action_space
         if isinstance(action_space, gym.spaces.Tuple):
             assert (
                 len(action_space.spaces) == 1
-            ), f"sb3 supports a single action space, this env constains multiple spaces {action_space}"
-
-    def step(self, action):
-        obs, reward, term, trunc, info = self.env.step(action)
-        obs = lod_to_dol(obs)
-
-        return {k: np.array(v) for k, v in obs.items()}, np.array(reward), np.array(term), info
-
-    def reset(self):
-        obs, info = self.env.reset()
-        obs = lod_to_dol(obs)
-        obs = {k: np.array(v) for k, v in obs.items()}
-        return obs
-
-    def close(self):
-        self.env.close()
-
-    def env_is_wrapped(self, wrapper_class, indices = None):
-        return [False] * self.env.num_envs
+            ), f"sb3 supports a single action space, this env contains multiple spaces {action_space}"
+
+    def step(self, action: np.ndarray) -> Tuple[Dict[str, np.ndarray], np.ndarray, np.ndarray, List[Dict[str, Any]]]:
+        # Initialize lists for collecting results
+        all_obs = []
+        all_rewards = []
+        all_term = []
+        all_trunc = []
+        all_info = []
+
+        # Get the number of environments
+        num_envs = self.envs[0].num_envs
+
+        # Send actions to each environment
+        for i in range(self.n_parallel):
+            self.envs[i].step_send(action[i*num_envs:(i+1)*num_envs])
+
+        # Receive results from each environment
+        for i in range(self.n_parallel):
+            obs, reward, term, trunc, info = self.envs[i].step_recv()
+            all_obs.extend(obs)
+            all_rewards.extend(reward)
+            all_term.extend(term)
+            all_trunc.extend(trunc)
```

Comment:
the truncated should be stored in info, and info should include terminal obs too, see https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/vec_env/dummy_vec_env.py#L61-L70

Reply:
@araffin Thanks for this info. I have a question. 

On the Godot side, envs are automatically reset when an episode terminates / truncates, I am not sure how feasible it is to send the terminal obs in the truncated setting without refactoring all the envs. Does the sb3 PPO implementation use the terminal obs for a value preduction in the truncated setting? If it is not possible to refactor in the short term, would you recommend just setting the terminated flag? (with trunc = False for all steps)